model-sync-gradle/src/main/kotlin/org/modelix/model/sync/gradle/config/ModelSyncGradleSettings.kt

```diff
@@ -0,0 +1,162 @@
+) {
+    fun fromModelServer(action: Action<ServerSource>) {
+        val endpoint = ServerSource()
+        action.execute(endpoint)
+        source = endpoint
+    }
+
+    fun fromLocal(action: Action<LocalSource>) {
+        val endpoint = LocalSource()
+        action.execute(endpoint)
+        source = endpoint
+    }
+
+    fun toModelServer(action: Action<ServerTarget>) {
+        val endpoint = ServerTarget()
+        action.execute(endpoint)
+        target = endpoint
+    }
+
+    fun toLocal(action: Action<LocalTarget>) {
+        val endpoint = LocalTarget()
+        action.execute(endpoint)
+        target = endpoint
+    }
+
+    fun includeModule(module: String) {
+        (includedModules as MutableSet).add(module)
+    }
+
+    fun registerLanguage(language: ILanguage) {
+        (registeredLanguages as MutableSet).add(language)
+    }
+}
+
+interface SyncEndPoint {
+    fun getValidationErrors(): String
+}
+
+sealed interface LocalEndpoint : SyncEndPoint {
+    var mpsHome: File?
+    var mpsHeapSize: String
+    var repositoryDir: File?
+
+    override fun getValidationErrors(): String {
+        return buildString {
+            if (mpsHome == null) {
+                appendUndefinedLocalFieldError("mpsHome")
+            }
+            if (repositoryDir == null) {
+                appendUndefinedLocalFieldError("repositoryDir")
+            }
+        }
+    }
+}
+
+data class LocalSource(
+    override var mpsHome: File? = null,
+    override var mpsHeapSize: String = "2g",
+    override var repositoryDir: File? = null,
+) : LocalEndpoint
+
+data class LocalTarget(
+    override var mpsHome: File? = null,
+    override var mpsHeapSize: String = "2g",
+    override var repositoryDir: File? = null,
+) : LocalEndpoint
+
+sealed interface ServerEndpoint : SyncEndPoint {
+    var url: String?
+    var repositoryId: String?
+    var branchName: String?
+
+    override fun getValidationErrors(): String {
+        return buildString {
+            if (url == null) {
+                appendUndefinedServerFieldError("url")
+            }
+        }
+    }
+}
+
+data class ServerSource(
+    override var url: String? = null,
+    override var repositoryId: String? = null,
+    override var branchName: String? = null,
+    var revision: String? = null,
+) : ServerEndpoint {
+    override fun getValidationErrors(): String {
+        return buildString {
+            append(super.getValidationErrors())
+            if (revision == null) {
+                if (repositoryId == null && branchName == null) {
+                    appendLine("Invalid server source. Please either specify a revision or repositoryId and branchName.")
+                } else if (repositoryId == null) {
+                    appendUndefinedServerFieldError("repositoryId")
+                } else if (branchName == null) {
+                    appendUndefinedServerFieldError("branchName")
+                }
+            }
```

Comment:
```suggestion
            if (revision != null) {
                // We do not need anything else if a revision is specified
                return
            }
            
            // Give users a proper hint in case the configuration is completely off
            if (revision == null && repositoryId == null && branchName == null) {
                    appendLine("Invalid server source. Please either specify a revision or repositoryId and branchName.")
                    return
            }
            
            // There are minor configuration issues.
            if (repositoryId == null) {
                    appendUndefinedServerFieldError("repositoryId")
            } else if (branchName == null) {
                    appendUndefinedServerFieldError("branchName")
            }
```

I had a hard time understanding the nested conditions. Maybe this version would be easier to understand?